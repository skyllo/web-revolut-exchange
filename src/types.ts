export enum Pocket {
  EUR = 'EUR',
  GBP = 'GBP',
  USD = 'USD',
}

export interface PocketData {
  name: Pocket;
  balance: number;
  symbol: string;
  title: string;
  countryCode: string;
}

export type InputDirection = 'from' | 'to';

export interface Rate {
  rate: number;
  fromPocket: Pocket;
  toPocket: Pocket;
}
