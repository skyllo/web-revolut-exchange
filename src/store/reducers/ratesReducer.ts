/* eslint-disable no-param-reassign */
import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';
import { getRate } from '../../api/client';
import { Pocket, Rate } from '../../types';

interface RateAction {
  fromPocket: Pocket;
  toPocket: Pocket;
}

export interface RateState {
  status: 'loading' | 'succeeded' | 'failed';
  error: string | undefined;
  data: Rate;
}

export const initialStateRate: RateState = {
  status: 'loading',
  error: undefined,
  data: {
    rate: 0,
    fromPocket: Pocket.GBP,
    toPocket: Pocket.EUR,
  },
};

export const fetchRate = createAsyncThunk<Rate, RateAction>(
  'rates/fetchRates',
  async ({ fromPocket, toPocket }) => getRate(fromPocket, toPocket)
);

export const ratesSlice = createSlice({
  name: 'rates',
  initialState: initialStateRate,
  reducers: {},
  extraReducers: (builder) => {
    builder.addCase(fetchRate.pending, (state) => {
      state.status = 'loading';
    });
    builder.addCase(fetchRate.fulfilled, (state, action) => {
      state.status = 'succeeded';
      state.data = action.payload;
    });
    builder.addCase(fetchRate.rejected, (state, action) => {
      state.status = 'failed';
      state.error = action.error.message;
    });
  },
});

export default ratesSlice.reducer;
