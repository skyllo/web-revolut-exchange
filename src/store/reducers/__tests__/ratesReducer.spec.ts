import { Pocket } from '../../../types';
import ratesReducer, { initialStateRate, fetchRate } from '../ratesReducer';

describe('ratesReducer', () => {
  it('should return the initial state', () => {
    expect(ratesReducer(undefined, { type: '' })).toMatchSnapshot();
  });

  it('should have correct state on successful rate', () => {
    const fromPocket = Pocket.EUR;
    const toPocket = Pocket.GBP;

    const action = fetchRate.fulfilled(
      {
        fromPocket,
        rate: 1.123123,
        toPocket,
      },
      '',
      { fromPocket, toPocket }
    );
    expect(ratesReducer(initialStateRate, action)).toMatchSnapshot();
  });

  it('should have correct state on failed rate', () => {
    const fromPocket = Pocket.EUR;
    const toPocket = Pocket.GBP;

    const action = fetchRate.rejected(new Error('failed!'), '', { fromPocket, toPocket });
    expect(ratesReducer(initialStateRate, action)).toMatchSnapshot();
  });

  it('should have correct state on loading rate', () => {
    const fromPocket = Pocket.EUR;
    const toPocket = Pocket.GBP;

    const action = fetchRate.pending('', { fromPocket, toPocket });
    expect(ratesReducer(initialStateRate, action)).toMatchSnapshot();
  });
});
