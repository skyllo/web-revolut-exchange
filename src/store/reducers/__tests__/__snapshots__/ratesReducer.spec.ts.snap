// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ratesReducer should have correct state on failed rate 1`] = `
Object {
  "data": Object {
    "fromPocket": "GBP",
    "rate": 0,
    "toPocket": "EUR",
  },
  "error": "failed!",
  "status": "failed",
}
`;

exports[`ratesReducer should have correct state on loading rate 1`] = `
Object {
  "data": Object {
    "fromPocket": "GBP",
    "rate": 0,
    "toPocket": "EUR",
  },
  "error": undefined,
  "status": "loading",
}
`;

exports[`ratesReducer should have correct state on successful rate 1`] = `
Object {
  "data": Object {
    "fromPocket": "EUR",
    "rate": 1.123123,
    "toPocket": "GBP",
  },
  "error": undefined,
  "status": "succeeded",
}
`;

exports[`ratesReducer should return the initial state 1`] = `
Object {
  "data": Object {
    "fromPocket": "GBP",
    "rate": 0,
    "toPocket": "EUR",
  },
  "error": undefined,
  "status": "loading",
}
`;
